### DISCUSSION TITLE
Potential Memory Leak in Internal server.

### DISCUSSION THREAD

**Discussion by SamLyuubc** [FreeOpcUa/opcua-asyncio#750] [Q&A]
### Background:
We recently upgrade to asyncua library to interact with ignition server OPCUA interface. We have a python application to host the opcua server to write/read to several nodes.  

### Issue
We have encountered a memory leak for sometimes and lately we caught the memory leak in a live debugging session. There are few findings as follow:

1. 93% of the heap memory is taken by list object (Shown in picture one)![overall_heap](https://user
![list_by_size](https://user-images.githubusercontent.com/23411719/145649143-c5150103-6171-4a3e-b2e6-718aebf89a39.png)
2. Most of the memory are taken by two lists, which consists 82% of the memory (Shown in picture two)![list_by_size](https://user-images.githubusercontent.com/23411719/145649183-550b28a3-def4-4f14-aa4a-d96ae1f9edeb.png)

3. Further more, the referrers of the biggest two lists are found (Shown in picture three):

- For the list with the largest size: Referrers by kind are dict of asyncua.ua.uaprotocol_auto.NotificationMessage and dict of asyncua.server.address_space.NodeData
- For the list with the second largest size: Referrers by kind are dict of asyncua.ua.uaportocol_auto.PublishResult and dict of asyncua.ua.uaprotocol_auto.DataChangeNotification![lists_byreference](https://user-images.githubusercontent.com/23411719/145649270-8f7f7398-21b3-4f48-a8ba-e9a42b76c0ed.png)

5. I picked one instance to analyze the reference pattern of the list, the result is shown in picture four
![referrence_pattern_for_the_largest_list](https://user-images.githubusercontent.com/23411719/145649293-6e9bd213-da61-474d-83c0-a6fd7b1e07a3.png)

###Question:
After digging into the source code, I suspect that internal server holds an large number of publish results and did not handle them properly. Did we implement the library wrong? What's the proper way to handle the publish results and avoid memory leak in this scenario?  I would appreciate your feedback on those questions. Thanks in advance.

**Comment by AndreasHeine** [comment]
could you provide more infos? 
the handling of the publish results is in internal_subscription.py -> https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/server/internal_subscription.py

**Reply by SamLyuubc** [reply]
Thanks for your comment Andrea. AFAIK, we just forward the result through opcua. But I will dig deeper and see if I can find anything useful.

**Comment by oroulet** [comment]
the most obvious reason would be if the client does not ackowledge our notfication requests for some resons. I think we are missing a safeguard here. if you trace your network with wireshark or similar you can see if the client acknowledge PublishResponses in their PublishRequest packages

**Reply by SamLyuubc** [reply]
@oroulet Thanks for your response. I did a tcpdump on the server(when there is no memory leak) and filter out the opcua messages. It looks like the publish request is including the subscription acknowledgement. I will try to capture packets when the memory leak happen again next time. 

For potential improvement, does it make sense for us to limit the number of publish results they are not sent properly so it does not use up all the memory? 


![publish_reponse request](https://user-images.githubusercontent.com/23411719/145877521-59c2be09-138f-4525-a53f-d36a2073a107.png)

**Reply by oroulet** [reply]
as far as I remember the list that may grow is that one: https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/server/internal_subscription.py#L37   I suppose the only correct behaviour is to terminate subscription if that list get higher than some number, not sure which one...

Also if you manage to find ut what list if growing, it would really help a lot to debug that issue

**Reply by SamLyuubc** [reply]
> 

Thanks @oroulet your suggestion is indeed correct. This week, I verified the list which grows continuously is the one:  https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/server/internal_subscription.py#L37

Currently, I am still working on the root cause why those results are not published successfully... There might be improper requests coming from other part of the systems. However, on the asyncua side, we would want to implement a safeguard for example, terminate the subscription if the list get higher than a threshold.

Also I noticed we mostly make use of the [pub_result_callback](https://github.com/FreeOpcUa/opcua-asyncio/blob/a17138c95c5002296745bdbd46696cbe823b4ea4/asyncua/server/internal_subscription.py#L115) function for communication through different system. Is it a option to set `no_acks` for the internal subscription over [here](https://github.com/FreeOpcUa/opcua-asyncio/blob/a17138c95c5002296745bdbd46696cbe823b4ea4/asyncua/server/internal_subscription.py#L21)?

**Reply by SamLyuubc** [reply]
Happy new year @oroulet . I did some experiment setting the `no_acks` parameter for `internal_sbuscription` to be False and the system works fine with no memory leak. However, I am reflecting the way we implement the datachange subscription and I think you might help me to get  a better understanding.

In our application, we have two servers(Let's name them A and B) and they also act as client to each other (See the attached figure below). The intention is to let server A subscribe to tags controlled by server B. To achieve this, in server A's software we use `server.create_subscription(250, handler)`  . Is this implementation problematic, since there are no clients sending publish request? Do I need to create a client instance in A's software and subscribing to the datachange?
![server client](https://user-images.githubusercontent.com/23411719/148609616-7087cb87-587c-4045-9c75-9ddf525c2fae.png)

**Reply by oroulet** [reply]
I am not 100% sure I follow. If you need your Server A applicatoin to get some data from Server B then you need a Client in your application too. This is no problem. An application can have many clients, on the other side, more than one server per application should not make sense in most cases.
so your drawing is a bit wrong it is more something like:

Application A         ----------------------   Application B
--- Server A           <---------------------------  Client to A
--- Client to B        --------------------------->  Server B

**Reply by SamLyuubc** [reply]
Thanks @oroulet  your answer is extremely helpful. What If I have one application A with server A and now I want to subscribe to the data change for some tags on server A. Do I need a separate client for subscription or I can just call the create_subscription function in server?
